[quant] Input-Weight Equalization - Conv prepare support

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61350
* #61287
* **#61286**
* #61285

Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557051](https://our.internmc.facebook.com/intern/diff/D29557051)

angelayi committed:
[quant] Input-Weight Equalization - Conv prepare support

Summary: Modifies the prepare step to support conv layers during input-weight equalization.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29557051).

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557051](https://our.internmc.facebook.com/intern/diff/D29557051)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557051](https://our.internmc.facebook.com/intern/diff/D29557051)

[ghstack-poisoned]

supriyar reviewed (approved):
test file changes look good

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557051](https://our.internmc.facebook.com/intern/diff/D29557051)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29557051).

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -211,13 +211,14 @@ def __new__(cls, input_activation=torch.nn.Identity, weight=torch.nn.Identity):
     if node.op == 'call_module':
         return (isinstance(modules[node.target], nn.Linear) or
-                isinstance(modules[node.target], nni.LinearReLU))
+                isinstance(modules[node.target], nni.LinearReLU) or
```
nit: we can use type(modules[node.target]) in [nn.Linear, nni.LinearReLU, nn.Conv2d]

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -211,13 +211,14 @@ def __new__(cls, input_activation=torch.nn.Identity, weight=torch.nn.Identity):
+                isinstance(modules[node.target], nn.Conv2d))
     elif node.op == 'call_function':
-        return node.target == F.linear
+        return node.target == F.linear or node.target == F.conv2d
```
same here

jerryzh168 reviewed (approved):
LGTM!

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557051](https://our.internmc.facebook.com/intern/diff/D29557051)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv prepare support"


Summary: Modifies the prepare step to support conv layers during input-weight equalization and adds tests to make sure that the results are as expected.

Initial: 
```
      w
      |
x -> conv -> y
```

After prepare:

```
                                         w
                                         |
                                  weight_quant_obs
                                         |
                                    weight_eq_obs
                                         |
x -> input_quant_obs -> input_eq_obs -> conv -> out_quant_obs -> y
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Initial:
```
ConvModel(
  (conv): Conv2d(3, 5, kernel_size=(3, 3), stride=(1, 1), bias=False)
)
```

After prepare:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_activation_post_process_0_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %conv : [#users=1] = call_module[target=conv](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %conv_activation_post_process_0 : [#users=1] = call_module[target=conv_activation_post_process_0](args = (%conv,), kwargs = {})
    return conv_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557051](https://our.internmc.facebook.com/intern/diff/D29557051)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29557051).
